### PR TITLE
Handle quoted scalars

### PR DESCRIPTION
**Pull request by rizo**
Fixes https://github.com/avsm/ocaml-yaml/issues/20.

I tried to find some information on implicit scalar quoting, but unfortunately it's not well documented. The only mention the spec has is for quoted implicit keys (see [Double-Quoted Style](https://yaml.org/spec/1.2/spec.html#id2787109) and [Single-Quoted Style](https://yaml.org/spec/1.2/spec.html#id2788097)). The only other explanation I found helpful was the "When not to use Plain Scalars" section [here](http://blogs.perl.org/users/tinita/2018/03/strings-in-yaml---to-quote-or-not-to-quote.html) (see the paragraph in the end about strings that would be resolved as a special types).

My conclusion from this is that quoted scalars must be treated as strings, which is what this patch does. I also tested a couple of other parsers and it seems they all do this.

If this looks good, I'll add some tests to cover the edge cases.

### CODE REVIEW

**Comment by avsm** [comment]
Thanks @rizo! The critical statement in your PR that leads me to merge this is:
>  I also tested a couple of other parsers and it seems they all do this.

Thanks very much for looking into that -- if you could add the test cases and yourself to the authors list in the opam file that'd be great. I'll move this repo to the mirage/ org and merge and add you as a collaborator if you like as well.

The CI failures are due to sexplib v0.12 being selected, so you can push a constraint to the opam file to stop that from happening too.

**Comment by avsm** [comment]
(you might notice I am a rather unopinionated maintainer when it comes to trying to the semantics of this library. Yaml seems like such a fast and loose format that the only viable strategy seems to be to mimic the behaviour of other implementations as much as is practical, like with Markdown before CommonMark showed up)

**Comment by rizo** [comment]
Thanks for the quick review, @avsm!

> I'll move this repo to the mirage/ org and merge and add you as a collaborator if you like as well.

That sounds good, I don't mind helping when needed.

> (...) the only viable strategy seems to be to mimic the behaviour of other implementations (...)

I think it's a sensible approach. We might inherit some issues from other implementations, but consistency with other parsers is probably more valuable for a data interchange format.

---

I added the tests as discussed. Additionally I decided to expose more information about the scalars in the `Yaml.yaml` type (see the `Yaml.scalar` type). I think it reflects better the spec and helps with scalar handling when converting to Json.

To help with tests I implemented an equality check function for Yaml values. It is exported in the public interface as it seems useful in general.

P.S. Sorry for all the whitespace changes, my editor automatically kills them. I think GitHub has an option to hide those.

File: lib/yaml.ml [hunk 1]
```diff
@@ -183,3 +189,16 @@ let pp ppf s =
   match to_string s with
   | Ok s -> Format.pp_print_string ppf s
   | Error (`Msg m) -> Format.pp_print_string ppf (Printf.sprintf "(error (%s))" m)
+
+let rec equal v1 v2 =
+  match v1, v2 with
+  | `Null, `Null -> true
+  | `Bool x1, `Bool x2 -> ((=) : bool -> bool -> bool) x1 x2
```
Comment (avsm) [inline lib/yaml.ml]: any reason not to write `x1 = x2` here?

Comment (avsm) [reply lib/yaml.ml]: Oh sorry, to specialise it -- `Bool x1` is a poly variant here so it wouldnt specialise based on the types of the pattern match

File: yaml.opam [hunk 2]
```diff
@@ -11,7 +11,7 @@ depends: [
   "dune" {build & >="1.3"}
   "ctypes" {>= "0.12.0"}
   "ppx_sexp_conv" {>= "v0.9.0"}
-  "sexplib"
+  "sexplib0" {< "v0.12"}
```
Comment (avsm) [inline yaml.opam]: I thinkkkkkkk the `< "v0.12"` isnt needed any more due to https://github.com/ocaml/opam-repository/pull/13674 being merged

**Comment by avsm** [comment]
All looks good. I'll merge when at a proper computer. Need to figure out why the Windows tests randomly broke :-/

**Comment by avsm** [comment]
Any objections to me applying ocamlformat to the repo btw? That'll sort out the whitespace issues you are seeing.

File: lib/yaml.ml (hunk 1 above)
Comment (rizo) [reply lib/yaml.ml]: Yeah, I haven't tested it, but I think specialising helps here. We could also use `[@@deriving eq]` to replace this implementation, but that would add an extra dependency.

File: yaml.opam (hunk 2 above)
Comment (rizo) [reply yaml.opam]: That's great! I'll remove this, hope it works.

**Comment by rizo** [comment]
> Any objections to me applying ocamlformat to the repo btw? That'll sort out the whitespace issues you are seeing.

No objections, feel free to! :)

> Need to figure out why the Windows tests randomly broke :-/

Anything I can help with? I'm not really familiar with the Windows ecosystem, but could take a look.

**Comment by rizo** [comment]
Seems like there are still some issues with sexplib, but it's not the inconsistent modules problem that we had previously. Should I add back the constraint or try to figure out the typing errors we see now?

**Comment by avsm** [comment]
Just throwing a quick appeal to @dra27 in case he can see what the mysterious Cygwin error is. It appears to be yet another new one!

Re: sexplib, putting the constraint back is indeed the way to go sadly.
```
File "types/bindings/yaml_bindings_types.ml", line 18, characters 26-31:
Error: This expression has type Sexplib0.Sexp.t
       but an expression was expected of type Sexplib.Type.t
```

cc @xclerc in case he knows if there is a way to remain compatible with both v0.11 and v0.12 of sexplib0...

**Comment by xclerc** [comment]
Sorry, I am a bit confused: the changes between v0.11 and v0.12 [1, 2]
seem to be related mainly to: (i) the build system, and (ii) the support for
macros (now in a dedicated library). I will have a closer look later today.

[1] https://github.com/janestreet/sexplib/compare/v0.11...v0.12?expand=1
[2] https://github.com/janestreet/sexplib0/compare/v0.11...v0.12?expand=1

**Comment by xclerc** [comment]
I was not able to reproduce the problem, but find the dependencies
of `ppx_sexp_conv.v0.11.0` quite suspicious. Indeed the package
does not declare any dependency on `sexplib0`.

In order to (in)validate this hypothesis, @avsm could you copy/paste
the output of `opam list` for the switch that produced the error above?

**Comment by rizo** [comment]
@xclerc I'm not sure if this is what you need, but here' what `opam install --deps-only -t --switch 4.06 .` installs in the CI (I believe it is equivalent to `opam list` you are asking for):

```
-> installed cmdliner.1.0.3
-> installed conf-m4.1
-> installed conf-pkg-config.1.1
-> installed ocamlfind.1.8.0
-> installed base-bytes.base
-> installed ocamlbuild.0.14.0
-> installed seq.0.1
-> installed uchar.0.0.2
-> installed num.1.1
-> installed dune.1.8.2
-> installed jbuilder.transition
-> installed cppo.1.6.5
-> installed ocaml-compiler-libs.v0.11.0
-> installed ppx_derivers.1.0
-> installed result.1.3
-> installed re.1.8.0
-> installed sexplib0.v0.11.0
-> installed ocaml-migrate-parsetree.1.2.0
-> installed topkg.1.0.0
-> installed astring.0.8.3
-> installed base.v0.11.1
-> installed fmt.0.8.5
-> installed fpath.0.7.2
-> installed integers.0.2.2
-> installed parsexp.v0.11.0
-> installed logs.0.6.2
-> installed ptime.0.8.4
-> installed rresult.0.6.0
-> installed mdx.1.3.0
-> installed sexplib.v0.11.0
-> installed cstruct.3.7.0
-> installed hex.1.3.0
-> installed stdio.v0.11.0
-> installed bos.0.2.0
-> installed uuidm.0.9.7
-> installed alcotest.0.8.5
-> installed uutf.1.0.2
-> installed jsonm.1.0.1
-> installed ezjsonm.1.0.0
-> installed tyxml.4.3.0
-> installed junit.2.0.1
-> installed junit_alcotest.2.0.1
-> installed ppxlib.0.5.0
-> installed ctypes.0.14.0
-> installed ppx_sexp_conv.v0.11.2
```

**Comment by xclerc** [comment]
Thanks @rizo! This is indeed the information I was looking
for, and it invalidates my hypothesis as all packages are at
the same version. It will try to reproduce the problem with
this set of packages.

**Comment by rizo** [comment]
I updated my post to include the actual output of `opam list` for 407.

If you wish to reproduce the environment in which the error occurs, you can use docker:

```
$ docker run -e AGENT_OS=LinuxDocker -e DISTRO=debian-stable -v `pwd`:/home/opam/src ocaml/opam2:debian-stable /home/opam/src/.azure-pipelines.sh
```

**Comment by dra27** [comment]
@avsm - you need a `.gitattributes` file as in https://github.com/mirage/mmap/pull/2

**Comment by rizo** [comment]
@avsm, I added back the constraint for sexplib0. The CI is passing now on all unix environments.

Regarding the Windows issue, @dra27's suggestion helped, but now the job is failing with a different error:

```
#=== ERROR while compiling conf-pkg-config.1.1 ================================#
# context              2.0.3 | win32/x86_64 | ocaml-variants.4.07.1+mingw64c | git+https://github.com/fdopen/opam-repository-mingw.git#opam2
# path                 ~/.opam/ocaml-variants.4.07.1+mingw64c/.opam-switch/build/conf-pkg-config.1.1
# command              D:\a\cygwin\home\VssAdministrator\.opam\ocaml-variants.4.07.1+mingw64c\bin\pkg-config.exe --help
# exit-code            2
# env-file             ~/.opam/log/conf-pkg-config-2840-0ae6fb.env
# output-file          ~/.opam/log/conf-pkg-config-2840-0ae6fb.out
### output ###
# pkg-config ist not installed. Install it with 'opam depext conf-pkg-config'
```

@dra27 do you know why `pkg-config` is not properly installed?

**Comment by avsm** [comment]
woohoo, we're green! (ignoring appveyor, i'll delete that)
